[마켓불리 김원희] Add: 늘 사는 것 Endpoint 구현 완료 (리뷰 반영하여 수정, 컨플릭트 수정)

supersaloon opened the pull request:
>wecode PR message template 및 체크 리스트 입니다. 
아래 사항들을 전부 작성/체크 하시고 PR 해주세요!

## 수정 사항 간략한 한줄 요약
- 늘 사는 것 end point 구현 완료
		 
## 수정 사항들 자세한 내용
- 늘 사는 것 상품 추가
  - 존재하지 상품 id이면 400 리턴
  - 이미 늘 사는 것에 존재하면 400 리턴
  - 이외의 경우 often_buyings 테이블에 데이터 추가
- 늘 사는 것 상품 조회
  - often_buyings 테이블에서 user_id 로 조회하여 해당되는 데이터들 출력
- 늘 사는 것 상품 삭제
  - often_buyings 테이블에서 상품 id 로 조회하여 해당되는 데이터 삭제

## 기타 질문 및 특이 사항
- 늘 사는 것(wish list) 관련하여 아래의 방식으로 접근하여 구현하여 보았습니다.
  - 아래와 같이 users 테이블에서 products 테이블에 often_buyings라는 중간 테이블을 걸쳐 ManyToMany 관계를 설정하였습니다. 
  - 늘 사는 것(wish list)을 사용자와 상품 사이에 있는 중간 테이블로 처리하였습니다.
  - 그리고 늘 사는 것(wish list)를 다루기 위해서 users나 products 테이블을 거치지 않고 직접 often_buyings라는 중간테이블에 접근하여, 거기서 products의 값들을 select_related로 불러와서 처리하였습니다.
- 이런식으로 접근하는 것이 일반적으로 E-commerce 서비스를 설계시에 wish list에 접근하는 방식인지, 그리고 더 좋은 방법이 있는지 궁금합니다.
```python
class User(models.Model):
    (생략)
    often_buying        = models.ManyToManyField(Product, through="OftenBuying", related_name='often_buying_set')
    (생략)
    class Meta:
        db_table = 'users'


class OftenBuying(models.Model):
    user     = models.ForeignKey("User", on_delete=models.CASCADE)
    product  = models.ForeignKey(Product, on_delete=models.SET_NULL, null=True)

    class Meta:
        db_table = 'often_buyings'


class Product(models.Model):
    name            = models.CharField(max_length=100)
    subtitle        = models.CharField(max_length=500)
    (생략)

    class Meta:
        db_table = 'products'
```

![자주 사는 것 테이블 질문](https://user-images.githubusercontent.com/69733417/102754400-4a9c6280-43b0-11eb-9ab1-643d2be45dad.PNG)



## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

Commit:
늘 사는 것 초기 세팅

Commit:
Add: 늘 사는 것에 상품 추가 기능 구현 완료
- 늘 사는 것에 상품 추가 요청을 받으면 해당 상품을 추가
- 존재하지 않는 상품에 대해서는 400 리턴
- 상품이 존재하더라도 이미 늘 사는 것에 추가되어 있으면 400 리턴

Commit:
Add: 늘 사는 것에 추가된 상품 조회
- 사용자가 늘 사는 것에 추가한 상품들 조회

Commit:
Add: 늘 사는 것에서 상품 삭제 구현 완료
- 요청 받은 늘사는 것 id에 대응하는 상품이 늘 사는 것에 있으면 삭제
- 늘 사는 것에 요청받은 id에 대응하는 상품이 없으면 400 리턴

soheon-lee commented:
네 원희님 질문 주신대로, Whishlist는 유저와 상품 사이의 중간 데이터로 다루는 것 좋은 방법입니다 :) 
[모델링 모음 사이트](http://www.databaseanswers.org/data_models/index.htm)
제가 헷갈리는 것이 있을 때마다 참고하는 사이트입니다. 다양한 모델링 케이스들이 많으니 한 번 참고해보세요!

soheon-lee commented on order/views.py:
```diff
@@ -1,3 +1,167 @@
+from django.http      import JsonResponse
+from django.views     import View
+
+from order.models   import Cart, Order
+from user.models      import Address, OftenBuying, User
+from product.models import Product
+
```
코드 간격 `import` 기준으로 정렬해주세요

supersaloon replied:
컨벤션에 맞도록 정렬하였습니다.

soheon-lee commented on order/views.py:
```diff
@@ -1,3 +1,167 @@
+                return JsonResponse({'MESSAGE': 'PRODUCT_DOES_NOT_EXIST'}, status=400)
+
+            if OftenBuying.objects.filter(user_id=user_id, product_id=data['product_id']).exists():
+                return JsonResponse({'MESSAGE': 'PRODUCT_ALREADY_EXIST_IN_OFTEN_BUYING'}, status=400)
+            else:
+                OftenBuying.objects.create(
+                    user_id    = user_id,
+                    product_id = data['product_id'],
+                )
+                return JsonResponse({'MESSAGE': 'SUCCESS'}, status=201)
```
else가 없어도 바로 위의 if문이 만족되지 않는다면 바로 도달할 수 있는 로직입니다. 꼭 필요한 경우가 아니면 else를 사용하지 않는 것이 보다 파이썬다운 코딩 스타일입니다. else는  지워주셔도 됩니다 :)

supersaloon replied:
불필요한 `else` 절 삭제하였습니다.

soheon-lee commented on order/views.py:
```diff
@@ -1,3 +1,167 @@
+    def get(self, request):  # 늘 사는 것 상품 조회
+        try:
+            user_id = 3  # 데코레이터 나오기 전까지 임의로 사용자 지정
+            often_buying = OftenBuying.objects.filter(user_id=user_id).select_related('product')
+
+            items_in_often_buying = [{
+                "id"               : item.id,
+                "product_id"       : item.product.id,
+                "name"             : item.product.name,
+                "price"            : item.product.price,
+                "image_url"        : item.product.image_url,
+            }for item in often_buying]
+
```
혹시 OftenBuying클래스에 직접 접근하여 user로 필터링 하는 방법 외에,
user_id를 기준으로 User 객체를 먼저 가져온 후, 연결된 oftenbuying 객체를 가져와서 (딱 그 사람의 위시리스트만 가져오도록) 해보는 방법을 시도해보시겠어요? ser객체에서 접근 하는 방법을 한 번 다뤄볼만한 미션인 것 같습니다 원희님 :)

supersaloon replied:
기존 often_buying 테이블에서 진입하는 방식에서 user 테이블에서 진입하는 방식으로 다시 구현해 보았습니다.

soheon-lee commented:
OftenBuying 윗부분은 order에서 리뷰 드린 내용이라 order merge되고 나서 고려하겠습니다 ~ :)

Commit:
Modify: PR #8 리뷰 반영하여 수정
- 컨벤션에 맞도록 코 드 수정
- 불필요한 `else` 구문 삭제
- 자주사는 것 목록을 받을 때 기존 OftenBuying 클래스에서 접근하는 방식이 아닌, User 클래스에서 접근하는 방식으로 구현

Commit:
Modify: conflic 수정
- market_bully/urls.py

Commit:
Modify: feature/order-cart 의 수정 내용 반영하여 view 파일 수정
- feature/order-cart와 feature/order-often_buyings 가 동일한 view 파일에 작성된 상태
- feature/order-cart에 수정사항 발생하여 수정 후 push하였고
- feature/order-often_buyings에서 수정된 내용을 동일하게 반영

supersaloon committed:
Merge branch 'master' into feature/order-often_buyings

jamessoun93 commented:
LGTM
